lsp/core/aws-lsp-core/src/credentials/ideCredentialsProvider.ts

```diff
@@ -0,0 +1,89 @@
+
+    constructor(private readonly connection: Connection, key?: string) {
+        if (key) {
+            this.key = Buffer.from(key, 'base64')
+            this.connection.console.info('Server: I was initialized with an encryption key')
+        } else {
+            this.connection.console.info("Server: I didn't get an encryption key. Functionality will be limited.")
+        }
+    }
+
+    /**
+     * Intended to be called when the language server is initialized.
+     * If the client provides credentials, handlers are registered.
+     */
+    public initialize(props: AwsInitializationOptions) {
+        this.registerCredentialsPushHandlers(props)
+    }
+
+    private async registerCredentialsPushHandlers(props: AwsInitializationOptions) {
+        if (props.credentials?.providesIam) {
+            this.registerIamCredentialsPushHandlers()
+        }
+
+        if (props.credentials?.providesBearerToken) {
+            this.connection.console.info('Server: (stub) Registering bearer token credentials push handlers')
+            // TODO : Set up Bearer token handlers
+        }
+    }
+
+    private registerIamCredentialsPushHandlers(): void {
+        this.connection.console.info('Server: Registering IAM credentials push handlers')
+
+        // Handle when host sends us credentials to use
+        this.connection.onNotification(
+            credentialsProtocolMethodNames.iamCredentialsUpdate,
+            (payload: UpdateCredentialsPayload) => {
+                const responseData = this.decryptResponseData(payload)
+                this.pushedCredentials = JSON.parse(responseData) as IamCredentials
+                this.connection.console.info('Server: The language server received updated credentials data.')
+            }
+        )
+
+        // Handle when host tells us we have no credentials to use
+        this.connection.onNotification(credentialsProtocolMethodNames.iamCredentialsClear, () => {
+            this.pushedCredentials = undefined
+            this.connection.console.info('Server: The language server does not have credentials anymore.')
+        })
+    }
+
+    /**
+     * Provides IAM based credentials. Throws NoCredentialsError if no credentials are available
+     */
+    public async resolveIamCredentials(token: CancellationToken): Promise<IamCredentials> {
+        if (!this.pushedCredentials) {
+            throw new NoCredentialsError()
+        }
+
+        return this.pushedCredentials
+    }
+
+    private decryptResponseData(response: UpdateCredentialsPayload): string {
+        if (!this.key) {
+            throw new Error('no encryption key')
+        }
+
+        const iv = Buffer.from(response.iv, 'base64')
+        const decipher = crypto.createDecipheriv('aes-256-gcm', this.key, iv, {
+            authTagLength: 16,
+        })
+        decipher.setAuthTag(Buffer.from(response.authTag, 'base64'))
+
+        return decipher.update(response.data, 'base64', 'utf8') + decipher.final('utf8')
```

Comment:
question: Does it make sense to group encipher and decipher functionality into a class of its own to keep the code close together and share details?

Reply:
I might have missed something -- this code is in a "core" package (aws-lsp-core) that can be consumed by multiple language server packages. It could be further separated from IdeCredentialsProvider, if we wanted to separate encryption/decryption code from the message handling code though.